### ISSUE TITLE
Audit & Fixes: Phishing Email Detection Project

### ISSUE THREAD

**Issue by NYN-05** [Click2Hack/Phishing-Email-Detection-Using-Machine-Learning#2]
**High Priority Issues**
- **Missing/incorrect requirements.txt location**:  
  - Problem: requirements.txt exists (not top-level). Typical tooling expects a root-level requirements.txt. Also the file contains unpinned package names (`Flask`, `pandas`, `scikit-learn`, `numpy`) without versions.  
  - Impact: Reproducibility and installs will vary and CI/automation may fail.  
  - Suggestion: Move or copy to repository root as requirements.txt, pin versions (e.g., `pandas==1.5.3`), and optionally add `requirements-dev.txt` for dev/test tools. Add instructions in README.md.  
  - Issue title: "Move and pin dependencies in root requirements.txt"

- **Project layout: app.py inside `requirements/` directory**:  
  - Problem: app.py contains the Flask app. This is confusing and non-standard.  
  - Impact: Developers will miss the web entrypoint; deployment tools will not find it.  
  - Suggestion: Move app.py to repository root or to `src/` (e.g., app.py) and update README and module imports. Update import path for `predict` if moved.  
  - Issue title: "Relocate app.py to top-level or `src/` and update imports"

- **No `.gitignore`** (not present in repo listing):  
  - Problem: Likely missing; models, venv, data, __pycache__ could be committed accidentally.  
  - Impact: Large/binary files committed (models, dataset), credential leakage, repo clutter.  
  - Suggestion: Add `.gitignore` including `venv/`, `__pycache__/`, `*.pyc`, `.env`, `/models/*` (except a `models/.gitkeep` if desired), `/data/*` if dataset is large or private.  
  - Issue title: "Add `.gitignore` to exclude venv, models, cache, and secrets"

- **Hard-coded relative paths and missing file checks**:  
  - Problem: Scripts use fixed paths e.g., `"data/phishing_emails.csv"`, `"data/preprocessed_data.pkl"`, `"models/phishing_detector.pkl"`. No robust path handling, no CLI args or environment config. `predict` expects a preprocessed pickle at `data/preprocessed_data.pkl`.  
  - Impact: Scripts fail when run from a different cwd or after restructuring. Harder integration into pipelines/CI.  
  - Suggestion: Add argument parsing (`argparse`), use `pathlib.Path(__file__).resolve().parent` or configuration via environment variables, and add user-friendly error messages when files missing.  
  - Issue title: "Add CLI args and robust path handling to scripts"

- **Lack of input validation and error handling in `predict` and Flask app**:  
  - Problem: app.py uses `request.json.get('email')` then passes to `predict_email` which will throw if model or vectorizer missing. `app.run(debug=True)` in production.  
  - Impact: Crash risk, information leak in production, possible remote code exposure via debug mode.  
  - Suggestion: Validate JSON structure, return clear 4xx/5xx messages, catch exceptions around model loading, disable Flask debug in production, and load model once at app startup rather than on each request. Use `werkzeug` error handling or Flask error handlers.  
  - Issue title: "Improve input validation and error handling in Flask app and toggle debug mode"

**Medium Priority Issues**
- **Model & vectorizer saving/loading approach is brittle**:  
  - Problem: preprocess.py pickle contains `(X, y, vectorizer)` saved to `data/preprocessed_data.pkl`, and train.py expects that. predict.py requires the same pickle as the vectorizer file. This couples preprocessing output and vectorizer storage to a data pickle. Model and vectorizer are saved separately (model in `models/`), but conventions vary.  
  - Impact: Confusion which files are required for runtime; large pickles may contain training arrays unnecessarily.  
  - Suggestion: Save vectorizer separately (e.g., `models/vectorizer.pkl`) and save processed features only if needed for experiments. Save model metadata (version, training date, hyperparams). Use joblib for sklearn artifacts for better performance. Add a small `load_model` helper.  
  - Issue title: "Save vectorizer and model artifacts separately; add metadata"

- **No tests or CI configuration**:  
  - Problem: There are no unit tests, and no CI (e.g., GitHub Actions) workflows.  
  - Impact: Hard to ensure changes don't break training/prediction; integration/deploy pipelines missing.  
  - Suggestion: Add unit tests for preprocess.py, train.py (small smoke test with a tiny dataset), and predict.py. Add GitHub Actions to run linting and tests on PRs.  
  - Issue title: "Add unit tests and GitHub Actions for CI"

- **README: formatting and inaccurate instructions**:  
  - Problem: Broken code block fences and incorrect virtualenv activation instructions for Windows (the README shows `venv\bin` and odd steps). It references requirements.txt in root but file sits in `requirements/`. Some code fences missing triple-backticks closure around `git clone` block.  
  - Impact: New users get confused and may not be able to run the project.  
  - Suggestion: Clean up README: fix code fences, provide exact PowerShell commands for Windows virtualenv activation (`.\venv\Scripts\Activate.ps1`), and provide example `python -m venv venv` + the correct `pip install -r requirements.txt` path. Document exact run commands for preprocessing, training, and serving.  
  - Issue title: "Fix README formatting and update usage instructions"

**Low Priority / Nice-to-have**
- **Add `.editorconfig` / linting / formatting**: add `flake8`, `black` config and a pre-commit hook.  
- **Add license check**: repository has a `LICENSE` but verify that authorship and license are intentional.  
- **Add CONTRIBUTING.md and CODE_OF_CONDUCT.md** for open-source friendliness.  
- **Remove stray files**: `stgfdhj` and `tempCodeRunnerFile.python` appear accidental; clarify and remove if not needed.  
  - Impact: Repo looks messy; accidental secrets risk.  
  - Issue title: "Remove or explain stray files (`stgfdhj`, `tempCodeRunnerFile.python`)"

**Security & Data Privacy**
- **Dataset licensing and privacy**:  
  - Problem: `data/phishing_emails.csv` included in repo; license and source not documented. May contain sensitive or copyrighted content.  
  - Suggestion: Document dataset licensing in README or remove dataset and point to an external download if licensing restricts redistribution. If dataset contains PII, remove it.  
  - Issue title: "Document data source and license; ensure no sensitive data is committed"

- **Running Flask with `debug=True`**:  
  - Problem: app.py calls `app.run(debug=True)` (unsafe in production).  
  - Suggestion: Use environment variable to control debug mode and recommend using a production WSGI server (gunicorn/uvicorn) for deployment.  
  - Issue title: "Disable Flask debug mode by default and recommend production deployment steps"

**Code-Quality Issues (concrete lines / behavior)**
- preprocess.py:
  - Uses `TfidfVectorizer().fit_transform(...).toarray()` — converting to dense array may be memory-heavy. Prefer keeping sparse matrices unless dense required.  
  - Saves `(X, y, vectorizer)` into a single pickle — see suggestions earlier.  
- train.py:
  - No hyperparameter control (all defaults). Consider exposing via arguments.  
  - No logging or seed control trade-offs beyond `random_state=42`.  
- predict.py:
  - Unclear contract for label encoding (assumes `1` == Phishing). Document label mapping consistently in preprocess.py / README.

**Suggested GitHub Issue Templates (copy-paste)**

- Title: Move and pin dependencies in root requirements.txt  
  Body:
  - Summary: Move requirements.txt to repository root as requirements.txt and pin dependency versions.  
  - Steps to reproduce: `pip install -r requirements.txt` fails because file missing in root.  
  - Proposed fix:
    - Create requirements.txt in repo root with pinned versions (example below).
    - Add `requirements-dev.txt` for dev tools.
    - Update README install instructions.
  - Example pinned content:
    ```
    pandas==1.5.3
    scikit-learn==1.2.2
    numpy==1.24.0
    Flask==2.2.5
    ```
  - Labels: `enhancement`, `dependencies`

- Title: Relocate `requirements/app.py` to top-level or `src/` and update imports  
  Body:
  - Summary: The Flask entrypoint is located at `requirements/app.py`, which is confusing. Move to `src/app.py` or root. Update `from src.predict import predict_email` import if file moved to root.  
  - Proposed fix:
    - Move file to `src/app.py`.
    - Update README run steps.
    - Load artifacts on start instead of per-request.

- Title: Add `.gitignore` to exclude venv, models, cache, and secrets  
  Body:
  - Add a `.gitignore` with common Python ignores and exclude `models/*`, unless you intend to commit model artifacts. Provide suggestions for allowing `models/.gitkeep`.  
  - Example:
    ```
    venv/
    .venv/
    __pycache__/
    *.pyc
    .env
    models/
    data/
    .ipynb_checkpoints/
    ```

- Title: Add unit tests and GitHub Actions for CI  
  Body:
  - Add `pytest` tests for `preprocess`, `train`, and `predict`. Add `.github/workflows/python-app.yml` to run tests on PRs.
